eth/downloader/downloader.go

```diff
@@ -1544,6 +1549,16 @@ func (d *Downloader) importBlockResults(results []*fetchResult) error {
 	if index, err := d.blockchain.InsertChain(blocks); err != nil {
 		if index < len(results) {
 			log.Debug("Downloaded item processing failed", "number", results[index].Header.Number, "hash", results[index].Header.Hash(), "err", err)
+
+			// In post-merge, notify the engine API of encountered bad chains
+			if d.badBlock != nil {
+				head, _, err := d.skeleton.Bounds()
+				if err != nil {
+					log.Error("Failed to retrieve beacon bounds for bad block reporting", "err", err)
+				} else {
+					d.badBlock(blocks[index].Header(), head)
```

Comment:
So blocks[index] is the bad block and head is the head that we currently trying to sync to, right?

Reply:
Yes

Reply:
This index can be inaccurate. e.g. here https://github.com/ethereum/go-ethereum/blob/master/core/blockchain.go#L1511 the returned index is the index of PrunedBlock, but not the real bad block position.

But this can be fixed in a following PR

Reply:
Recovering the ancestor shouldn't fail though, it if was already once imported?

Reply:
theoretically, yes. It shouldn't fail.